zerver/worker/queue_processors.py

```diff
@@ -527,3 +534,43 @@ def consume(self, event: Mapping[str, Any]) -> None:
                 (stream, recipient, sub) = access_stream_by_id(user_profile, stream_id,
                                                                require_active=False)
                 do_mark_stream_messages_as_read(user_profile, stream)
+
+        elif event['type'] == 'export_user_messages':
+            user_profile = get_user_profile_by_id(event['user_profile_id'])
+
+            # Create export
+            output_dir = tempfile.mkdtemp(prefix="/tmp/zulip-export-")
+            do_export_user(user_profile, output_dir)
+            print("Finished exporting to %s; tarring" % (output_dir,))
+            tarball_path = output_dir.rstrip('/') + '.tar.gz'
+            subprocess.check_call(
+                ["tar", "-czf", tarball_path, os.path.basename(output_dir)],
+                cwd=os.path.dirname(output_dir)
+            )
+            print("Tarball written to %s" % (tarball_path,))
+
+            # Upload
+            with open(tarball_path, 'rb') as f:
+                content_type = mimetypes.guess_type(tarball_path)[0]
+                size = os.path.getsize(tarball_path)
+                name = os.path.basename(tarball_path)
+                print(content_type, name, size)
+                upload_path = upload_message_file(name, size, content_type, f.read(), user_profile)
+
+            # Delete export data
+            shutil.rmtree(output_dir)
+            os.remove(tarball_path)
+
+            # Notify user
+            content = """Your data has been exported and uploaded [here]({})""".format(upload_path)
+            internal_send_private_message(user_profile.realm,
+                                          get_system_bot(settings.NOTIFICATION_BOT),
+                                          user_profile,
+                                          content)
+
+    def consume_wrapper(self, data: Dict[str, Any]) -> None:
+        """Over-ridden to mark user data export as finished even on failures."""
+        super(DeferredWorker, self).consume_wrapper(data)
+        if data['type'] == 'export_user_messages':
+            user_id = data['user_profile_id']
+            UserDataExport.export_finished(user_id)
```

Comment:
This might be easier done as a `try/except` block in the handler after extracting the function.